### ISSUE TITLE
Missing keyboard input functionality

### ISSUE THREAD

**Issue by ThomasMiz** [Ultz/Silk.NET#399]
There currently doesn't appear to be a way to properly detect some keys for keyboard typing input.

The Keyboard.KeyChar provides the characters typed, and keys such as Enter, Backspace, Tab, Delete can be detected via KeyDown. However, if the user is pressing down on the backspace key, the only thing we can register with this system is a single backspace press, and no repeats.

Looking at [the GLFW input guide](https://www.glfw.org/docs/latest/input_guide.html#input_key), there aren't just "key down" and "key up" actions, but there's also a "key repeat" action which we're probably ignoring.

We should try to present all these parameters in our KeyChar, KeyDown and KeyUp events.

**Comment by SanFlan** [comment]
I'm not that familiarized with SILK but it seems that the event is being purposely ignored in GlfwKeyboard.cs, line 49

            _handle = events.Handle;
            events.Char += _char = (_, c) => KeyChar?.Invoke(this, (char) c);
            events.Key += _key = (_, key, code, action, mods) =>
                (action switch
                {
                    InputAction.Press => KeyDown,
                    InputAction.Release => KeyUp,
                    InputAction.Repeat => null,
                    _ => null
                })?.Invoke(this, ConvertKey(key), code);`

**Comment by ThomasMiz** [comment]
I think it's a good idea to propose some changes to the input events so we can pass all these parameters to the user of the library.

We'll have to add more parameters to the KeyDown & other events, which will mean everyone will have to add them to their event handlers as well. It might be a good idea to do as OpenTK and aggregate all these parameters into a single struct called KeyEventArgs or something like that, which will be the sole parameter of these events:

```cs
public readonly struct KeyEventArgs
{
    public Key Key { get; }
    public int KeyCode { get; }
    public bool IsRepeat { get; }
    public ModifierKeys ModifiersKeys { get; }
    // And any other param which we might want to add now or in the future
}
```
This way we can also add more parameters in the future without making breaking changes.

We can invoke the "repeated" keys with the KeyDown event.

The other option would be to replace KeyDown and KeyUp with a single key event which reports all three up down and repeat, in which case instead of a `bool IsRepeat` the struct would have a `KeyAction Action { get; }`, but I like the previous approach more.

**Comment by HurricanKai** [comment]
Sounds like a good idea to me - but not something we can do right now, this would have to be 3.0 work, which is faaar off.
For now, we could simply expose a new event Repeat that triggers when a repeat is triggered.
This isn't ideal from the user code side, but at least works.

**Comment by Perksey** [comment]
We could just have a catch-all Key event.
```cs
event Action<KeyEvent> Key;
```

**Comment by ThomasMiz** [comment]
Just adding the Key event is an easy solution, but not a good solution.

What I suggest is a breaking change but previous code will be easy to adapt- All you have to do is change the event handler's parameters to a single one and then get the parameters from there. And it solves all these problems.

I think we should make these changes, if not for Silk 2.X then for Silk 3.0 and optionally, until then, we can add the Key event Perksey suggested.

**Comment by Perksey** [comment]
We can't do breaking changes for 2.X so it'll have to be 3.X.

Silk.NET 3.0 has an official release date of December 31st 9999 but realistically won't be any time before 2022 (I'm not dedicating any time to new massive features for 3.0, only minor improvements, bug fixes, and the monthly updates).

**Comment by FrostByteGER** [comment]
> 
> 
> We can't do breaking changes for 2.X so it'll have to be 3.0.
> 
> Silk.NET 3.0 has an official release date of December 31st 9999 but realistically won't be any time before 2022 (I'm not dedicating any time to new massive features for 3.0, only minor improvements, bug fixes, and the monthly updates).

So 3.0 will be more like a big maintenance release?

**Comment by ThomasMiz** [comment]
I created a proposal for these changes. You can find it [here](https://github.com/ThomasMiz/Silk.NET/blob/feature/input-event-enhancements/documentation/proposals/Proposal%20-%20Enhanced%20Input%20Events.md).

**Comment by Perksey** [comment]
> So 3.0 will be more like a big maintenance release?

@FrostByteGER 3.0 has a tracker here: #209. It'll be added to as we further realise 3.0 but we haven't even started thinking about 3.0 yet.

**Comment by Perksey** [comment]
@ThomasMiz if you can pull request this in like #397 we'll review it next working group meeting.

**Comment by MatijaBrown** [comment]
> There currently doesn't appear to be a way to properly detect some keys for keyboard typing input.
> 
> The Keyboard.KeyChar provides the characters typed, and keys such as Enter, Backspace, Tab, Delete can be detected via KeyDown. However, if the user is pressing down on the backspace key, the only thing we can register with this system is a single backspace press, and no repeats.
> 
> Looking at [the GLFW input guide](https://www.glfw.org/docs/latest/input_guide.html#input_key), there aren't just "key down" and "key up" actions, but there's also a "key repeat" action which we're probably ignoring.
> 
> We should try to present all these parameters in our KeyChar, KeyDown and KeyUp events.

If you want to do text input, you should use the char callback not the keydown event, according to GLFW. This also automatically handles key modifiers like AltGr+Q=@ and so on. Doesn't fix the backspace problem though.

**Comment by Perksey** [comment]
This has been resolved in the 3.0 proposals #539

**Comment by Pannoniae** [comment]
btw this is still a problem

**Comment by Beyley** [comment]
> btw this is still a problem

if you have an issue with the solution provided by the 3.0 proposals, please open a separate issue regarding any missing features from there

**Comment by Pannoniae** [comment]
no, I mean it's an issue right now, not in 3.0 when trying to use the library:) I'm literally repackaging Silk.NET.Input to add this functionality rn

**Comment by Perksey** [comment]
Yes we’re aware it’s an issue right now, however the work specific to this issue is deemed complete for 3.0. Everything else will be tackled as part of normal 3.0 development. Very similar to how we treat issues in 2.X where we don’t close the issue until the work specific to the issue is in a core branch. We don’t keep it open until shipment because that would create a lot of maintenance overhead.

**Comment by Pannoniae** [comment]
Is it not possible to just add a repeat event as a kludge, like right now? That won't break anything in terms of backward compatibility.

**Comment by Perksey** [comment]
Any modification of the API contract will be breaking.

**Comment by Pannoniae** [comment]
How does adding an event break the API? I don't understand

**Comment by Perksey** [comment]
Any third party implementation of IKeyboard would be incompatible with the update unless it also adjusts its code.
